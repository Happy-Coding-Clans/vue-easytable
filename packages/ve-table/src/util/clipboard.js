const regUniversalNewLine = /^(\r\n|\n\r|\r|\n)/;
const regNextCellNoQuotes = /^[^\t\r\n]+/;
const regNextEmptyCell = /^\t/;

/**
 * @decodeSpreadsheetStr
 * @desc Decode spreadsheet string into array.  refer from http://github.com/warpech/sheetclip/
 * @param {string} str The string to parse.
 * @returns {array}
 */
export function decodeSpreadsheetStr(str) {
    let arr = [[""]];

    if (str.length === 0) {
        return arr;
    }

    let column = 0;
    let row = 0;
    let lastLength;

    while (str.length > 0) {
        if (lastLength === str.length) {
            // In the case If in last cycle we didn't match anything, we have to leave the infinite loop
            break;
        }

        lastLength = str.length;

        if (str.match(regNextEmptyCell)) {
            str = str.replace(regNextEmptyCell, "");

            column += 1;
            arr[row][column] = "";
        } else if (str.match(regUniversalNewLine)) {
            str = str.replace(regUniversalNewLine, "");
            column = 0;
            row += 1;

            arr[row] = [""];
        } else {
            let nextCell = "";

            if (str.startsWith('"')) {
                let quoteNo = 0;
                let isStillCell = true;

                while (isStillCell) {
                    const nextChar = str.slice(0, 1);

                    if (nextChar === '"') {
                        quoteNo += 1;
                    }

                    nextCell += nextChar;

                    str = str.slice(1);

                    if (
                        str.length === 0 ||
                        (str.match(/^[\t\r\n]/) && quoteNo % 2 === 0)
                    ) {
                        isStillCell = false;
                    }
                }

                nextCell = nextCell
                    .replace(/^"/, "")
                    .replace(/"$/, "")
                    .replace(/["]*/g, (match) =>
                        new Array(Math.floor(match.length / 2))
                            .fill('"')
                            .join(""),
                    );
            } else {
                const matchedText = str.match(regNextCellNoQuotes);

                nextCell = matchedText ? matchedText[0] : "";
                str = str.slice(nextCell.length);
            }

            arr[row][column] = nextCell;
        }
    }
    // 去除 excel 最后一个多余的换行数据
    if (Array.isArray(arr) && arr.length > 1) {
        if (arr[arr.length - 1].length === 1 && arr[arr.length - 1][0] === "") {
            arr = arr.slice(0, arr.length - 1);
        }
    }

    return arr;
}

/**
 * @decodeSpreadsheetStr
 * @desc encode array to spreadsheet string.  refer from http://github.com/warpech/sheetclip/
 * @param {array} str The string to parse.
 * @returns {string}
 */
export function encodeToSpreadsheetStr(arr) {
    let r;
    let rLen;
    let c;
    let cLen;
    let str = "";
    let val;

    for (r = 0, rLen = arr.length; r < rLen; r += 1) {
        cLen = arr[r].length;

        for (c = 0; c < cLen; c += 1) {
            if (c > 0) {
                str += "\t";
            }
            val = arr[r][c];

            if (typeof val === "string") {
                if (val.indexOf("\n") > -1) {
                    str += `"${val.replace(/"/g, '""')}"`;
                } else {
                    str += val;
                }
            } else if (val === null || val === void 0) {
                // void 0 resolves to undefined
                str += "";
            } else {
                str += val;
            }
        }

        if (r !== rLen - 1) {
            str += "\n";
        }
    }

    return str;
}

/**
 * @onBeforeCopy
 * @desc on before copy
 * @param {Event} event
 * @return {selectionRangeIndexes,selectionRangeKeys,data}
 */
export function onBeforeCopy({
    cellSelectionRangeData,
    selectionRangeData,
    colgroups,
    allRowKeys,
}) {
    const { leftColKey, rightColKey, topRowKey, bottomRowKey } =
        cellSelectionRangeData;

    const selectionRangeIndexes = {
        startColIndex: colgroups.findIndex((x) => x.key === leftColKey),
        endColIndex: colgroups.findIndex((x) => x.key === rightColKey),
        startRowIndex: allRowKeys.indexOf(topRowKey),
        endRowIndex: allRowKeys.indexOf(bottomRowKey),
    };

    const selectionRangeKeys = {
        startColKey: leftColKey,
        endColKey: rightColKey,
        startRowKey: topRowKey,
        endRowKey: bottomRowKey,
    };

    const response = {
        selectionRangeIndexes,
        selectionRangeKeys,
        data: selectionRangeData,
    };

    return response;
}

/**
 * @onAfterCopy
 * @desc on after copy
 * @param {Event} event
 * @return
 */
export function onAfterCopy({ event, selectionRangeData }) {
    const spreadsheetStr = encodeToSpreadsheetStr(selectionRangeData);

    if (event.clipboardData) {
        event.clipboardData.setData("text/plain", spreadsheetStr);
    }
    // IE browser
    else if (window.clipboardData) {
        window.clipboardData.setData("Text", spreadsheetStr);
    }
}

/**
 * @onBeforePaste
 * @desc on before paste
 * @param {Event} event
 * @return
 */
export function onBeforePaste({
    event,
    cellSelectionRangeData,
    colgroups,
    allRowKeys,
}) {
    let pastedData;

    if (event.clipboardData) {
        pastedData = event.clipboardData.getData("text/plain");
    }
    // IE browser
    else if (window.clipboardData) {
        pastedData = window.clipboardData.getData("Text");
    }

    if (typeof pastedData !== "string") {
        return null;
    }

    let decodePasteData = decodeSpreadsheetStr(pastedData);

    const startColIndex = colgroups.findIndex(
        (x) => x.key === cellSelectionRangeData.leftColKey,
    );
    const endColIndex = Math.min(
        startColIndex + decodePasteData[0].length - 1,
        colgroups.length - 1,
    );
    const startRowIndex = allRowKeys.indexOf(cellSelectionRangeData.topRowKey);
    const endRowIndex = Math.min(
        startRowIndex + decodePasteData.length - 1,
        allRowKeys.length - 1,
    );

    let response = {
        selectionRangeIndexes: {
            startColIndex,
            endColIndex,
            startRowIndex,
            endRowIndex,
        },
        selectionRangeKeys: {
            startColKey: colgroups[startColIndex].key,
            endColKey: colgroups[endColIndex].key,
            startRowKey: allRowKeys[startRowIndex],
            endRowKey: allRowKeys[endRowIndex],
        },
        data: [],
    };

    const sourceFieldNames = colgroups
        .slice(startColIndex, endColIndex + 1)
        .map((x) => x.field);

    response.data = decodePasteData
        .slice(0, endRowIndex - startRowIndex + 1)
        .map((rowData) => {
            let newRow = {};

            rowData.forEach((cellData, cellIndex) => {
                if (cellIndex <= endColIndex - startColIndex) {
                    newRow[sourceFieldNames[cellIndex]] = cellData;
                }
            });
            return newRow;
        });

    return response;
}

/**
 * @onAfterPaste
 * @desc on after paste
 * @param {Event} event
 * @return
 */
export function onAfterPaste({ tableData, beforePasteResponse }) {
    const { data: pasteData, selectionRangeIndexes } = beforePasteResponse;

    pasteData.forEach((rowData, rowIndex) => {
        Object.assign(
            tableData[selectionRangeIndexes.startRowIndex + rowIndex],
            rowData,
        );
    });
}

/**
 * @onBeforeCut
 * @desc on before cut
 * @param {Event} event
 * @return {selectionRangeIndexes,selectionRangeKeys,data}
 */
export function onBeforeCut({
    cellSelectionRangeData,
    selectionRangeData,
    colgroups,
    allRowKeys,
}) {
    const { leftColKey, rightColKey, topRowKey, bottomRowKey } =
        cellSelectionRangeData;

    const selectionRangeIndexes = {
        startColIndex: colgroups.findIndex((x) => x.key === leftColKey),
        endColIndex: colgroups.findIndex((x) => x.key === rightColKey),
        startRowIndex: allRowKeys.indexOf(topRowKey),
        endRowIndex: allRowKeys.indexOf(bottomRowKey),
    };

    const selectionRangeKeys = {
        startColKey: leftColKey,
        endColKey: rightColKey,
        startRowKey: topRowKey,
        endRowKey: bottomRowKey,
    };

    const response = {
        selectionRangeIndexes,
        selectionRangeKeys,
        data: selectionRangeData,
    };

    return response;
}

/**
 * @onAfterCut
 * @desc on after cut
 * @param {Event} event
 * @return
 */
export function onAfterCut({
    event,
    tableData,
    colgroups,
    selectionRangeData,
    selectionRangeIndexes,
}) {
    const spreadsheetStr = encodeToSpreadsheetStr(selectionRangeData);

    const { endColIndex, endRowIndex, startColIndex, startRowIndex } =
        selectionRangeIndexes;

    // 移除制定的表格数据
    const fieldNames = colgroups
        .slice(startColIndex, endColIndex + 1)
        .map((x) => x.field);

    tableData.forEach((rowData, rowIndex) => {
        if (rowIndex >= startRowIndex && rowIndex <= endRowIndex) {
            fieldNames.forEach((fieldName) => {
                rowData[fieldName] = "";
            });
        }
    });

    if (event.clipboardData) {
        event.clipboardData.setData("text/plain", spreadsheetStr);
    }
    // IE browser
    else if (window.clipboardData) {
        window.clipboardData.setData("Text", spreadsheetStr);
    }
}

/**
 * @onBeforeDelete
 * @desc on before delete
 * @param {Event} event
 * @return {selectionRangeIndexes,selectionRangeKeys,data}
 */
export function onBeforeDelete({
    cellSelectionRangeData,
    selectionRangeData,
    colgroups,
    allRowKeys,
}) {
    const { leftColKey, rightColKey, topRowKey, bottomRowKey } =
        cellSelectionRangeData;

    const selectionRangeIndexes = {
        startColIndex: colgroups.findIndex((x) => x.key === leftColKey),
        endColIndex: colgroups.findIndex((x) => x.key === rightColKey),
        startRowIndex: allRowKeys.indexOf(topRowKey),
        endRowIndex: allRowKeys.indexOf(bottomRowKey),
    };

    const selectionRangeKeys = {
        startColKey: leftColKey,
        endColKey: rightColKey,
        startRowKey: topRowKey,
        endRowKey: bottomRowKey,
    };

    const response = {
        selectionRangeIndexes,
        selectionRangeKeys,
        data: selectionRangeData,
    };

    return response;
}

/**
 * @onAfterDelete
 * @desc on after delete
 * @param {Event} event
 * @return
 */
export function onAfterDelete({ tableData, colgroups, selectionRangeIndexes }) {
    const { endColIndex, endRowIndex, startColIndex, startRowIndex } =
        selectionRangeIndexes;

    // 移除制定的表格数据
    const fieldNames = colgroups
        .slice(startColIndex, endColIndex + 1)
        .map((x) => x.field);

    tableData.forEach((rowData, rowIndex) => {
        if (rowIndex >= startRowIndex && rowIndex <= endRowIndex) {
            fieldNames.forEach((fieldName) => {
                rowData[fieldName] = "";
            });
        }
    });
}
