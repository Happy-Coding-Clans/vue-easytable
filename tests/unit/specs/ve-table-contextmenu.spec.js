import { mount } from "@vue/test-utils";
import veTable from "@/ve-table";
import { later } from "../util";
import { cloneDeep } from "lodash";

describe("veTable contextmenu", () => {
    const TABLE_DATA = [
        {
            name: "John",
            date: "1900-05-20",
            number: "32",
            address: "No.1 Century Avenue, Shanghai",
            rowKey: 0,
        },
        {
            name: "Dickerson",
            date: "1910-06-20",
            number: "676",
            address: "No.1 Century Avenue, Beijing",
            rowKey: 1,
        },
        {
            name: "Larsen",
            date: "2000-07-20",
            number: "76",
            address: "No.1 Century Avenue, Chongqing",
            rowKey: 2,
        },
        {
            name: "Geneva",
            date: "2010-08-20",
            number: "7797",
            address: "No.1 Century Avenue, Xiamen",
            rowKey: 3,
        },
        {
            name: "Jami",
            date: "2020-09-20",
            number: "8978",
            address: "No.1 Century Avenue, Shenzhen",
            rowKey: 4,
        },
    ];

    const COLUMNS = [
        {
            field: "",
            key: "a",
            title: "",
            width: 50,
            align: "center",
            renderBodyCell: ({ row, column, rowIndex }, h) => {
                return ++rowIndex;
            },
        },
        {
            field: "name",
            key: "name",
            title: "Name",
            align: "left",
            width: "15%",
        },
        {
            field: "date",
            key: "date",
            title: "Date",
            align: "left",
            width: "15%",
        },
        {
            field: "number",
            key: "number",
            title: "Number",
            align: "right",
            width: "30%",
        },
        {
            field: "address",
            key: "address",
            title: "Address",
            align: "left",
            width: "40%",
        },
    ];

    const CONTEXTMENUS = [
        {
            type: "INSERT_ROW_ABOVE",
        },
        {
            type: "INSERT_ROW_BELOW",
        },
        {
            type: "SEPARATOR",
        },
        {
            type: "REMOVE_ROW",
        },
        {
            type: "HIDE_COLUMN",
        },
    ];

    afterEach(() => {
        return new Promise((resolve) => {
            const contextmenuPopper = document.querySelector(
                ".ve-contextmenu-popper",
            );
            contextmenuPopper && contextmenuPopper.remove();

            resolve();
        });
    });

    it("render", () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: COLUMNS,
                tableData: TABLE_DATA,
                rowKeyFieldName: "rowKey",
                contextmenuBodyOption: {
                    callback: ({ type, selection }) => {
                        console.log("type::", type);
                        console.log("selection::", selection);
                    },
                    contextmenus: CONTEXTMENUS,
                },
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it("contextmenu INSERT_ROW_ABOVE", async () => {
        const mockFn = jest.fn();

        let tableData = cloneDeep(TABLE_DATA);

        const WRAPPER = mount(
            {
                render() {
                    return (
                        <div>
                            <ve-table
                                row-key-field-name="rowKey"
                                columns={COLUMNS}
                                table-data={tableData}
                                contextmenu-body-option={
                                    this.contextmenuBodyOption
                                }
                            />
                        </div>
                    );
                },
                data() {
                    return {
                        contextmenuBodyOption: {
                            callback: ({ type, selection }) => {
                                mockFn({ type, selection });
                            },
                            contextmenus: CONTEXTMENUS,
                        },
                    };
                },
            },
            // need attach to documnet
            { attachTo: document.body },
        );

        const firstTrTdEl = WRAPPER.findAll(".ve-table-body-tr")
            .at(2)
            .findAll(".ve-table-body-td")
            .at(2);

        firstTrTdEl.trigger("mousedown");

        await later();

        expect(firstTrTdEl.classes()).toContain("ve-table-cell-selection");

        const bodyEl = WRAPPER.find(".ve-table-body");
        bodyEl.trigger("contextmenu");

        await later();

        const contextmenuPopper = document.querySelector(
            ".ve-contextmenu-popper",
        );

        const contextmenuNodes = contextmenuPopper.querySelectorAll(
            ".ve-contextmenu-node",
        );

        const event2 = new MouseEvent("click", {
            view: window, // window
            bubbles: true,
            cancelable: true,
        });

        contextmenuNodes[0].dispatchEvent(event2);

        await later();

        expect(tableData.length).toBe(6);

        expect(mockFn).toHaveBeenCalled();
        expect(mockFn).toHaveBeenCalledWith({
            selection: { colKey: "date", rowKey: 2, rowIndex: 2 },
            type: "INSERT_ROW_ABOVE",
        });
    });

    it("contextmenu INSERT_ROW_BELOW", async () => {
        const mockFn = jest.fn();

        let tableData = cloneDeep(TABLE_DATA);

        const WRAPPER = mount(
            {
                render() {
                    return (
                        <div>
                            <ve-table
                                row-key-field-name="rowKey"
                                columns={COLUMNS}
                                table-data={tableData}
                                contextmenu-body-option={
                                    this.contextmenuBodyOption
                                }
                            />
                        </div>
                    );
                },
                data() {
                    return {
                        contextmenuBodyOption: {
                            callback: ({ type, selection }) => {
                                mockFn({ type, selection });
                            },
                            contextmenus: CONTEXTMENUS,
                        },
                    };
                },
            },
            // need attach to documnet
            { attachTo: document.body },
        );

        const firstTrTdEl = WRAPPER.findAll(".ve-table-body-tr")
            .at(2)
            .findAll(".ve-table-body-td")
            .at(2);

        firstTrTdEl.trigger("mousedown");

        await later();

        expect(firstTrTdEl.classes()).toContain("ve-table-cell-selection");

        const bodyEl = WRAPPER.find(".ve-table-body");
        bodyEl.trigger("contextmenu");

        await later();

        const contextmenuPopper = document.querySelector(
            ".ve-contextmenu-popper",
        );

        const contextmenuNodes = contextmenuPopper.querySelectorAll(
            ".ve-contextmenu-node",
        );

        const event2 = new MouseEvent("click", {
            view: window, // window
            bubbles: true,
            cancelable: true,
        });

        contextmenuNodes[1].dispatchEvent(event2);

        await later();

        expect(tableData.length).toBe(6);

        expect(mockFn).toHaveBeenCalled();
        expect(mockFn).toHaveBeenCalledWith({
            selection: { colKey: "date", rowKey: 2, rowIndex: 2 },
            type: "INSERT_ROW_BELOW",
        });
    });

    it("contextmenu REMOVE_ROW", async () => {
        const mockFn = jest.fn();

        let tableData = cloneDeep(TABLE_DATA);

        const WRAPPER = mount(
            {
                render() {
                    return (
                        <div>
                            <ve-table
                                row-key-field-name="rowKey"
                                columns={COLUMNS}
                                table-data={tableData}
                                contextmenu-body-option={
                                    this.contextmenuBodyOption
                                }
                            />
                        </div>
                    );
                },
                data() {
                    return {
                        contextmenuBodyOption: {
                            callback: ({ type, selection }) => {
                                mockFn({ type, selection });
                            },
                            contextmenus: CONTEXTMENUS,
                        },
                    };
                },
            },
            // need attach to documnet
            { attachTo: document.body },
        );

        const firstTrTdEl = WRAPPER.findAll(".ve-table-body-tr")
            .at(2)
            .findAll(".ve-table-body-td")
            .at(2);

        firstTrTdEl.trigger("mousedown");

        await later();

        expect(firstTrTdEl.classes()).toContain("ve-table-cell-selection");

        const bodyEl = WRAPPER.find(".ve-table-body");
        bodyEl.trigger("contextmenu");

        await later();

        const contextmenuPopper = document.querySelector(
            ".ve-contextmenu-popper",
        );

        const contextmenuNodes = contextmenuPopper.querySelectorAll(
            ".ve-contextmenu-node",
        );

        const event2 = new MouseEvent("click", {
            view: window, // window
            bubbles: true,
            cancelable: true,
        });

        contextmenuNodes[2].dispatchEvent(event2);

        await later();

        expect(tableData.length).toBe(4);

        expect(mockFn).toHaveBeenCalled();
        expect(mockFn).toHaveBeenCalledWith({
            selection: { colKey: "date", rowKey: 2, rowIndex: 2 },
            type: "REMOVE_ROW",
        });
    });

    it("contextmenu HIDE_COLUMN", async () => {
        const mockFn = jest.fn();

        let tableData = cloneDeep(TABLE_DATA);

        const WRAPPER = mount(
            {
                render() {
                    return (
                        <div>
                            <ve-table
                                row-key-field-name="rowKey"
                                columns={COLUMNS}
                                table-data={tableData}
                                contextmenu-body-option={
                                    this.contextmenuBodyOption
                                }
                            />
                        </div>
                    );
                },
                data() {
                    return {
                        contextmenuBodyOption: {
                            callback: ({ type, selection }) => {
                                mockFn({ type, selection });
                            },
                            contextmenus: CONTEXTMENUS,
                        },
                    };
                },
            },
            // need attach to documnet
            { attachTo: document.body },
        );

        const firstTrTdEl = WRAPPER.findAll(".ve-table-body-tr")
            .at(2)
            .findAll(".ve-table-body-td")
            .at(2);

        firstTrTdEl.trigger("mousedown");

        await later();

        expect(firstTrTdEl.classes()).toContain("ve-table-cell-selection");

        const bodyEl = WRAPPER.find(".ve-table-body");
        bodyEl.trigger("contextmenu");

        await later();

        const DateTh = WRAPPER.findAll(
            ".ve-table-header-tr .ve-table-header-th",
        ).at(2);
        expect(DateTh.text()).toBe("Date");

        const contextmenuPopper = document.querySelector(
            ".ve-contextmenu-popper",
        );

        const contextmenuNodes = contextmenuPopper.querySelectorAll(
            ".ve-contextmenu-node",
        );

        const event2 = new MouseEvent("click", {
            view: window, // window
            bubbles: true,
            cancelable: true,
        });

        contextmenuNodes[3].dispatchEvent(event2);

        await later();

        const DateTh2 = WRAPPER.findAll(
            ".ve-table-header-tr .ve-table-header-th",
        ).at(2);
        expect(DateTh2.text()).toBe("Number");

        expect(mockFn).toHaveBeenCalled();
        expect(mockFn).toHaveBeenCalledWith({
            selection: { colKey: "date", rowKey: 2, rowIndex: 2 },
            type: "HIDE_COLUMN",
        });
    });
});
