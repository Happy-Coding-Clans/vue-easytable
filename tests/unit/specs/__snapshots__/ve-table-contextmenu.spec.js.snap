// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`veTable contextmenu render 1`] = `
<div class="ve-table ve-table-border-around">
  <div class="ve-table-container ve-table-enable-cell-selection" style="height: 0px;">
    <div class="ve-table-content-wrapper">
      <table class="ve-table-content ve-table-border-x">
        <colgroup>
          <col style="width: 50px;">
          <col style="width: 15%;">
          <col style="width: 15%;">
          <col style="width: 30%;">
          <col style="width: 40%;">
        </colgroup>
        <thead class="ve-table-fixed-header ve-table-header">
          <tr class="ve-table-header-tr">
            <th rowspan="1" colspan="1" class="ve-table-header-th" style="text-align: center; top: 0px;"></th>
            <th rowspan="1" colspan="1" class="ve-table-header-th" style="text-align: left; top: 0px;">Name</th>
            <th rowspan="1" colspan="1" class="ve-table-header-th" style="text-align: left; top: 0px;">Date</th>
            <th rowspan="1" colspan="1" class="ve-table-header-th" style="text-align: right; top: 0px;">Number</th>
            <th rowspan="1" colspan="1" class="ve-table-header-th ve-table-last-column" style="text-align: left; top: 0px;">Address</th>
          </tr>
        </thead>
        <tbody class="ve-table-body ve-table-row-hover ve-table-row-highlight">
          <tr style="height: 0px;">
            <td style="padding: 0px; border: 0px; height: 0px;"></td>
            <td style="padding: 0px; border: 0px; height: 0px;"></td>
            <td style="padding: 0px; border: 0px; height: 0px;"></td>
            <td style="padding: 0px; border: 0px; height: 0px;"></td>
            <td style="padding: 0px; border: 0px; height: 0px;"></td>
          </tr>
          <tr row-key="0" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="a" class="ve-table-body-td" style="text-align: center;">1</td>
            <td rowspan="1" colspan="1" col-key="name" class="ve-table-body-td" style="text-align: left;">John</td>
            <td rowspan="1" colspan="1" col-key="date" class="ve-table-body-td" style="text-align: left;">1900-05-20</td>
            <td rowspan="1" colspan="1" col-key="number" class="ve-table-body-td" style="text-align: right;">32</td>
            <td rowspan="1" colspan="1" col-key="address" class="ve-table-body-td" style="text-align: left;">No.1 Century Avenue, Shanghai</td>
          </tr>
          <tr row-key="1" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="a" class="ve-table-body-td" style="text-align: center;">2</td>
            <td rowspan="1" colspan="1" col-key="name" class="ve-table-body-td" style="text-align: left;">Dickerson</td>
            <td rowspan="1" colspan="1" col-key="date" class="ve-table-body-td" style="text-align: left;">1910-06-20</td>
            <td rowspan="1" colspan="1" col-key="number" class="ve-table-body-td" style="text-align: right;">676</td>
            <td rowspan="1" colspan="1" col-key="address" class="ve-table-body-td" style="text-align: left;">No.1 Century Avenue, Beijing</td>
          </tr>
          <tr row-key="2" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="a" class="ve-table-body-td" style="text-align: center;">3</td>
            <td rowspan="1" colspan="1" col-key="name" class="ve-table-body-td" style="text-align: left;">Larsen</td>
            <td rowspan="1" colspan="1" col-key="date" class="ve-table-body-td" style="text-align: left;">2000-07-20</td>
            <td rowspan="1" colspan="1" col-key="number" class="ve-table-body-td" style="text-align: right;">76</td>
            <td rowspan="1" colspan="1" col-key="address" class="ve-table-body-td" style="text-align: left;">No.1 Century Avenue, Chongqing</td>
          </tr>
          <tr row-key="3" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="a" class="ve-table-body-td" style="text-align: center;">4</td>
            <td rowspan="1" colspan="1" col-key="name" class="ve-table-body-td" style="text-align: left;">Geneva</td>
            <td rowspan="1" colspan="1" col-key="date" class="ve-table-body-td" style="text-align: left;">2010-08-20</td>
            <td rowspan="1" colspan="1" col-key="number" class="ve-table-body-td" style="text-align: right;">7797</td>
            <td rowspan="1" colspan="1" col-key="address" class="ve-table-body-td" style="text-align: left;">No.1 Century Avenue, Xiamen</td>
          </tr>
          <tr row-key="4" class="ve-table-body-tr">
            <td rowspan="1" colspan="1" col-key="a" class="ve-table-body-td" style="text-align: center;">5</td>
            <td rowspan="1" colspan="1" col-key="name" class="ve-table-body-td" style="text-align: left;">Jami</td>
            <td rowspan="1" colspan="1" col-key="date" class="ve-table-body-td" style="text-align: left;">2020-09-20</td>
            <td rowspan="1" colspan="1" col-key="number" class="ve-table-body-td" style="text-align: right;">8978</td>
            <td rowspan="1" colspan="1" col-key="address" class="ve-table-body-td" style="text-align: left;">No.1 Century Avenue, Shenzhen</td>
          </tr>
        </tbody>
        <tfoot class="ve-table-fixed-footer ve-table-footer"></tfoot>
      </table>
      <div class="ve-table-selection-wrapper">
        <div class="ve-table-selection-fixed-left"></div>
        <div class="ve-table-selection-middle"></div>
        <div class="ve-table-selection-fixed-right"></div>
      </div>
    </div>
  </div>
  <div class="ve-table-edit-input-container" style="top: 0px; left: 0px; height: 1px; z-index: -1; opacity: 0;"><textarea tabindex="-1" class="ve-table-edit-input"></textarea></div>
  <div class="ve-contextmenu" style="display: none;">
    <div class="ve-contextmenu-panel">
      <ul class="ve-contextmenu-list">
        <li class="ve-contextmenu-node"><span class="ve-contextmenu-node-label">insert row above</span></li>
        <li class="ve-contextmenu-node"><span class="ve-contextmenu-node-label">insert row below</span></li>
        <li class="ve-contextmenu-node-separator"></li>
        <li class="ve-contextmenu-node"><span class="ve-contextmenu-node-label">remove row</span></li>
        <li class="ve-contextmenu-node"><span class="ve-contextmenu-node-label">hide column</span></li>
      </ul>
    </div>
  </div>
</div>
`;
